File: frontend2/lib/models/alert_model.dart
class AlertModel {
  final String id;
  final String title;
  final String body;
  final bool hasCountdown;
  final int expiresAt;
  final String targetType;

  AlertModel({
    required this.id,
    required this.title,
    required this.body,
    required this.hasCountdown,
    required this.expiresAt,
    required this.targetType,
  });

  factory AlertModel.fromJson(Map<String, dynamic> json) {
    return AlertModel(
      id: json['id'] ?? '',
      title: json['title'] ?? '',
      body: json['body'] ?? '',
      // Handle string booleans from FCM payloads
      hasCountdown: json['hasCountdown'] == "true" || json['hasCountdown'] == true,
      // Handle string ints from FCM payloads
      expiresAt: int.tryParse(json['expiresAt'].toString()) ?? 0,
      targetType: json['targetType'] ?? 'global',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'body': body,
      'hasCountdown': hasCountdown ? "true" : "false",
      'expiresAt': expiresAt.toString(),
      'targetType': targetType,
    };
  }
}

File: server/v1/modules/alert/alertModel.js
// alert/alertModel.js

const mongoose = require("mongoose");

const alertSchema = new mongoose.Schema(
  {
    title: { type: String, required: true },
    body: { type: String, required: true },
    hasCountdown: { type: Boolean, default: false },
    expiresAt: { type: Date, required: true },
    targetType: { 
      type: String, 
      enum: ["mess", "hostel", "global"], 
      required: true 
    },
    // References to Hostel IDs (used for both hostels and messes in your system)
    targetIds: [{ type: mongoose.Schema.Types.ObjectId, ref: "Hostel" }],
    createdBy: { type: mongoose.Schema.Types.ObjectId, ref: "User" }
  },
  { timestamps: true }
);

// Automatically delete documents from MongoDB when expiresAt is reached
alertSchema.index({ expiresAt: 1 }, { expireAfterSeconds: 0 });
// Optimize read queries
alertSchema.index({ targetType: 1, targetIds: 1 });

module.exports = mongoose.model("Alert", alertSchema);
